File: Java/src/StreamPrinter.java
import java.io.IOException;
import java.io.Reader;

public class StreamPrinter {

    void print(Reader reader) throws IOException {
        if (reader == null) {
            throw new IllegalArgumentException("Reader cannot be null");
        }
        try (Reader r = reader) {
            int code;
            while ((code = r.read()) != -1) {
                System.out.println((char) code);
            }
        }
    }
}

File: Java/kotlin/StreamPrinter.kt
package kotlin

import java.io.IOException
import java.io.Reader

class StreamPrinter {
    @Throws(IOException::class)
    fun print(reader: Reader?) {
        requireNotNull(reader) { "Reader cannot be null" }

        reader.use { r ->
            var code: Int
            while (r.read().also { code = it } != -1) {
                println(code.toChar())
            }
        }
    }
}

